File: 0834-sum-of-distances-in-tree/0834-sum-of-distances-in-tree.py
class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        graph = defaultdict(set)
        for u, v in edges: graph[u].add(v), graph[v].add(u)

        def dfs1(u, l):
            if u in seen: return
            seen.add(u)
            total[0] += l
            for v in graph[u]: dfs1(v, l + 1) 

        def dfs2(u):
            if u in seen: return 0
            seen.add(u)
            ans = 1 + sum(dfs2(v) for v in graph[u])
            counts[u] = ans
            return ans

        def dp(u, parent):
            if u in seen: return
            seen.add(u)
            if u != 0: ans[u] = ans[parent] + n - 2 * counts[u] 
            for v in graph[u]: dp(v, u)

        total, seen = [0], set()
        dfs1(0, 0)
        counts, seen = [0] * n, set()
        dfs2(0)
        ans, seen = [total[0]] * n, set()
        dp(0, 0)
        return ans

File: 0834-sum-of-distances-in-tree/0834-sum-of-distances-in-tree.cpp
class Solution {
public:
    int sum=0;
    int helper(vector<int> adj[],int node,vector<int> &vis,vector<int> &numEdges,int dis){
        int ret=0;
        sum+=dis;
        vis[node]=1;
        for(auto a: adj[node]){
            if(!vis[a]){
                ret+=1+helper(adj,a,vis,numEdges,dis+1);
                numEdges[node]=ret;
            }
        }
        return ret;
    }
    void dfs(vector<int> adj[],int node,int n,vector<int>& numEdges,vector<int>& vis,vector<int>& res){
        vis[node]=1;
        for(auto a: adj[node]){
            if(!vis[a]){
                res[a]=res[node]-numEdges[a]+n-2-numEdges[a];
                dfs(adj,a,n,numEdges,vis,res);
            }
        }
    }
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        vector<int> adj[n];
        for(int i=0;i<edges.size();i++){
            adj[edges[i][0]].push_back(edges[i][1]);
            adj[edges[i][1]].push_back(edges[i][0]);
        }
        vector<int> numEdges(n,0);
        vector<int> vis(n,0);
        vector<int> res(n,0);
        helper(adj,0,vis,numEdges,0);
        res[0]=sum;
        vector<int> vis1(n,0);
        dfs(adj,0,n,numEdges,vis1,res);
        return res;
    }
};